include/nbl/builtin/hlsl/cmath.hlsl

```diff
@@ -0,0 +1,286 @@
+}
+
+
+// Exponential and logarithmic functions
+NBL_ALIAS_UNARY_FUNCTION(exp)
+NBL_ALIAS_FUNCTION_WITH_OUTPUT_PARAM(frexp, int32_t)
+NBL_ALIAS_BINARY_FUNCTION(ldexp)
+NBL_ALIAS_UNARY_FUNCTION(log)
+NBL_ALIAS_UNARY_FUNCTION(log10)
+NBL_ALIAS_FUNCTION_WITH_OUTPUT_PARAM(modf, T)
+NBL_ALIAS_UNARY_FUNCTION(exp2)
+NBL_ALIAS_UNARY_FUNCTION(log2)
+NBL_ALIAS_UNARY_FUNCTION2(logb,log)
+
+template<class T> 
+T expm1(T x) 
+{ 
+    //wrong
+    // for better implementation, missing function for evaluate_polynomial
+    // https://www.boost.org/doc/libs/1_83_0/boost/math/special_functions/expm1.hpp
+
+    //or use std::expm1f, std::expm1l depending on size of T
+    return INTRINSIC_NAMESPACE(exp)(x) - T(1); 
+}
+
+template<class T> 
+T log1p(T x) 
+{ 
+    // https://www.boost.org/doc/libs/1_83_0/boost/math/special_functions/log1p.hpp
+    /*if (x < T(-1))
+         "log1p(x) requires x > -1, but got x = %1%.
+    if (x == T(-1))
+        return -inf */
+    T u = T(1) + x;
+    if (u == T(1))
+        return x;
+    else
+        return INTRINSIC_NAMESPACE(log)(u) * (x / (u - T(1)));
+}
+
+template<class T>
+int32_t ilogb(T x) 
+{ 
+    using uint_type = typename nbl::hlsl::numeric_limits<T>::uint_type;
+    const int32_t shift = (impl::num_base<T>::float_digits-1);
+    const uint_type mask = ~(((uint_type(1) << shift) - 1) | (uint_type(1)<<(sizeof(T)*8-1)));
+    int32_t bits = (std::bit_cast<uint_type, T>(x) & mask) >> shift; //nbl::hlsl::bit_cast sucks as it pulls the ambiguous one from intrinsics.h
+    return bits + impl::num_base<T>::min_exponent - 2;
+}
+
+template<class T> 
+T scalbn(T x, int32_t n) 
+{ 
+    return x * INTRINSIC_NAMESPACE(exp2)(n); 
+}
+
+// Power functions
+NBL_ALIAS_BINARY_FUNCTION(pow)
+NBL_ALIAS_UNARY_FUNCTION(sqrt)
+NBL_ALIAS_UNARY_FUNCTION(cbrt)
+
+template<class T>
+T hypot(T x, T y)
+{
+    return INTRINSIC_NAMESPACE(sqrt)(x*x+y*y);
+}
+
+// Floating-point manipulation functions
+
+template<class T>
+T copysign(T x, T sign_)
+{
+    using uint_type = uint32_t; //typename nbl::hlsl::numeric_limits<T>::uint_type;
+    constexpr uint_type m = (1u << 31);
+    const uint_type a = std::bit_cast<uint_type, T>(sign_) & m;
+    const uint_type b = std::bit_cast<uint_type, T>(x) & (m-1);
+    return std::bit_cast<T, uint_type>(a ^ b);
+}
+
+// Generate quiet NaN (function)
+template<class T>
+T nan()
+{
+    return nbl::core::nan<T>();
+}
+// TODO:
+// nextafter	Next representable value (function) 
+// nexttoward	Next representable value toward precise value (function)
+
+// Error and gamma functions
+
+template<class T>
+T erf(T x)
+{
+    // Bürmann series approximation  
+    // https://www.desmos.com/calculator/myf9ylguh1
+    // https://en.wikipedia.org/wiki/Error_function#Numerical_approximations
+    T E = INTRINSIC_NAMESPACE(exp)(-x*x);
+    T P = T(0.886226925453);
+    T re = INTRINSIC_NAMESPACE(sqrt)(T(1)-E)*(T(1)+T(.155)*E/P*(T(1)-T(.275)*E));
+    return INTRINSIC_NAMESPACE(copysign)(re, x);
+}
```

Comment:
Remove all code/impl thats either subpar or we don't have a license/software-bill-of-materials for